src/Utils/normalization.jl

```diff
@@ -1,68 +1,119 @@
 
-### Output
-
-Either the same system if the inputs are of type `AbstractInput`, or a new
-system that wraps the inputs in a `ConstantInput`.
+If the system does not conform to a canonical form, the implementation tries
+to make the transformation; otherwise an error is thrown. In particular, ODEs
+of the form ``x' = Ax + Bu`` are mapped into ``x' = Ax + u, u ∈ B\\mathcal{U}``,
+where now ``u`` has the same dimensions as ``x``.
 """
-function normalize_inputs(system)
-    inputdim(system) == 0 && return system
-    U = inputset(system)
-    U isa AbstractInput && return system
-    if !(U isa LazySet)
-        throw(ArgumentError("inputs of type $(typeof(U)) are not supported"))
-    end
-    return IVP(wrap_inputs(system.s, U), system.x0)
+function normalize(system::AbstractSystem)
+    throw(ArgumentError("the system type $(typeof(system)) is currently not supported"))
 end
 
-function wrap_inputs(system::CLCCS, U::LazySet)
-    return CLCCS(system.A, system.B, system.X, ConstantInput(U))
+# initial value problems in canonical form, i.e. which don't need normalization
+normalize(system::CF) = system
+
+# x' = Ax + Bu, x ∈ X, u ∈ U
+function normalize(system::CLCCS{N, AN, BN, XT, UT}) where {N, AN<:AbstractMatrix{N}, BN<:AbstractMatrix{N}, XT<:XNCF, UT<:UNCF}
+    n = statedim(system)
+    X = _wrap_invariant(system.X, n)
+    U = _wrap_inputs(system.U, system.B)
+    CLCCS(system.A, I(n, N), X, U)
 end
 
-function wrap_inputs(system::CLCDS, U::LazySet)
-    return CLCDS(system.A, system.B, system.X, ConstantInput(U))
+# x' = Ax + Bu, x ∈ X, u ∈ U
+function normalize(system::CLCDS{N, AN, BN, XT, UT}) where {N, AN<:AbstractMatrix{N}, BN<:AbstractMatrix{N}, XT<:XNCF, UT<:UNCF}
+    n = statedim(system)
+    X = _wrap_invariant(system.X, n)
+    U = _wrap_inputs(system.U, system.B)
+    CLCDS(system.A, I(n, N), X, U)
 end
 
-function wrap_inputs(system::CACCS, U::LazySet)
-    return CACCS(system.A, system.B, system.c, system.X, ConstantInput(U))
+# x' = Ax + Bu + c, x ∈ X, u ∈ U
+function normalize(system::CACCS{N, AN, BN, VN, XT, UT}) where {N, AN<:AbstractMatrix{N}, BN<:AbstractMatrix{N}, VN<:AbstractVector{N}, XT<:XNCF, UT<:UNCF}
+    n = statedim(system)
+    X = _wrap_invariant(system.X, n)
+    U = _wrap_inputs(system.U, system.B, system.c)
+    CACCS(system.A, I(n, N), X, U)
 end
 
-function wrap_inputs(system::CACDS, U::LazySet)
-    return CACDS(system.A, system.B, system.c, system.X, ConstantInput(U))
+function normalize(system::CACDS{N, AN, BN, VN, XT, UT}) where {N, AN<:AbstractMatrix{N}, BN<:AbstractMatrix{N}, VN<:AbstractVector{N}, XT<:XNCF, UT<:UNCF}
+    n = statedim(system)
+    X = _wrap_invariant(system.X, n)
+    U = _wrap_inputs(system.U, system.B, system.c)
+    CACDS(system.A, I(n, N), X, U)
 end
 
+_wrap_invariant(X::LazySet, n::Int) = X
+_wrap_invariant(X::Nothing, n::Int) = Universe(n)
+
+_wrap_inputs(U::AbstractInput, B::IdentityMultiple) = U
+_wrap_inputs(U::AbstractInput, B::AbstractMatrix) = map(u -> B*u, U)
```

Comment:
What is the result type of `map`?

Reply:
The result type of `map` is the same as that of `U`, see [`map(f::Function, c::ConstantInput)`](https://github.com/JuliaReach/MathematicalSystems.jl/blob/master/src/inputs.jl#L77) and [`map(f::Function, v::VaryingInput)`](https://github.com/JuliaReach/MathematicalSystems.jl/blob/master/src/inputs.jl#L199).